File: VendingMachineApp.py
# Smart Vending Machine (User Flow) - Python Console
# Features:
#  - Browse items with price & stock
#  - Add to Cart (validates stock)
#  - Checkout (cash only). Can add more cash or cancel.
#  - Dispense simulation + Receipt printing
#  - Stock auto-updates after purchase
# How to run:  python vending_machine.py

from collections import OrderedDict
from datetime import datetime
import uuid

class Item:
    def __init__(self, name: str, price: int, stock: int):
        self.name = name
        self.price = price  # kyats
        self.stock = stock

class Cart:
    def __init__(self):
        # keep insertion order for pretty receipt
        self.lines: "OrderedDict[Item,int]" = OrderedDict()

    def add(self, item: Item, qty: int):
        self.lines[item] = self.lines.get(item, 0) + qty

    def total(self) -> int:
        return sum(i.price * q for i, q in self.lines.items())

    def is_empty(self) -> bool:
        return len(self.lines) == 0

    def clear(self):
        self.lines.clear()

class VendingMachine:
    MACHINE_ID = "VM-102"
    LOCATION   = "Cafeteria Building A"

    def __init__(self):
        self.inventory: "OrderedDict[int, Item]" = OrderedDict()
        self.cart = Cart()
        self.seed_inventory()

    # ---------- Inventory & UI ----------
    def seed_inventory(self):
        # Drinks
        self.inventory[1] = Item("Coca-Cola", 2000, 10)
        self.inventory[2] = Item("Pepsi",     1800, 10)
        self.inventory[3] = Item("Sprite",    1500, 10)
        self.inventory[4] = Item("Water",     1000, 10)
        # Snacks
        self.inventory[5] = Item("Lays Chips",     1500, 8)
        self.inventory[6] = Item("Cookies",        1200, 6)
        self.inventory[7] = Item("Nutter Butter",  1300, 5)
        self.inventory[8] = Item("Kit Kat",        1500, 7)

    def show_menu(self):
        print("\n=========== Available Items ===========")
        print(f"{'No.':<4} {'Name':<18} {'Price(Ks)':>10} {'Stock':>7}")
        for no, it in self.inventory.items():
            print(f"{no:<4} {it.name:<18} {it.price:>10,} {it.stock:>7}")
        print("=======================================")
        if not self.cart.is_empty():
            print(f"Cart Total: {self.cart.total():,} Ks")
        print("(Type item number to add, C=Checkout, Q=Quit)")

    @staticmethod
    def ask_int(prompt: str, min_v: int, max_v: int) -> int:
        while True:
            s = input(prompt).strip()
            try:
                n = int(s)
                if n < min_v or n > max_v:
                    print(f"Please enter a number between {min_v} and {max_v}.")
                    continue
                return n
            except ValueError:
                print("Invalid number. Try again.")

    # ---------- Main Loop ----------
    def run(self):
        print("Welcome to Smart Vending Machine")
        print("Cash Payment Only | Type 'CANCEL' during payment to abort")

        while True:
            self.show_menu()
            choice = input("\nChoose: [number=add / C=Checkout / Q=Quit] : ").strip()

            if choice.lower() == "q":
                print("\nGoodbye! Thanks for visiting.")
                break

            if choice.lower() == "c":
                if self.cart.is_empty():
                    print("Cart is empty. Please add at least one item.")
                    continue
                if self.checkout():
                    again = input("\nBuy another item? (Y/N): ").strip()
                    if again.lower() != "y":
                        print("\nThank you! See you again.")
                        break
                continue

            # add to cart
            try:
                num = int(choice)
            except ValueError:
                print("Invalid choice. Try again.")
                continue

            item = self.inventory.get(num)
            if not item:
                print("Invalid selection number.")
                continue
            if item.stock <= 0:
                print(f"Out of stock: {item.name}")
                continue

            qty = self.ask_int(f"Enter quantity for {item.name}: ", 1, item.stock)
            self.cart.add(item, qty)
            print(f"Added to cart: {item.name} x{qty} ({item.price:,} Ks each)")
            input("Press Enter to continue...")

    # ---------- Checkout & Payment ----------
    def checkout(self) -> bool:
        print("\n================= CHECKOUT =================")
        self.print_cart()
        total = self.cart.total()
        print(f"TOTAL: {total:,} Ks")

        ok = input("Proceed to payment? (Y/N): ").strip()
        if ok.lower() != "y":
            print("Returning to menu...")
            return False

        paid = 0
        while paid < total:
            s = input(f"Insert cash (current paid {paid:,} / total {total:,}): ").strip()
            if s.lower() == "cancel":
                print("Transaction canceled. Returning to main menu.")
                return False
            try:
                add = int(s)
                if add <= 0:
                    print("Please insert a positive amount (or type CANCEL).")
                    continue
            except ValueError:
                print("Invalid amount. Type a number or 'CANCEL'.")
                continue

            paid += add
            if paid < total:
                print(f"Insufficient funds. Need {(total - paid):,} Ks more.")

        change = paid - total

        # Final stock validation
        for it, q in self.cart.lines.items():
            if it.stock < q:
                print(f"Sorry, '{it.name}' stock changed. Only {it.stock} left. Canceling.")
                return False

        # Dispense + stock update
        print("\nDispensing items...")
        for it, q in self.cart.lines.items():
            it.stock -= q
            print(f" - {it.name} x{q}")

        # Print receipt
        self.print_receipt(paid, change)

        # Clear cart
        self.cart.clear()
        return True

    def print_cart(self):
        print("-------------------------------------------")
        for it, q in self.cart.lines.items():
            line_total = it.price * q
            print(f"{it.name:<16} x{q:<3} {line_total:>10,} Ks")
        print("-------------------------------------------")

    def print_receipt(self, paid: int, change: int):
        total = sum(it.price * q for it, q in self.cart.lines.items())
        txn_id = str(uuid.uuid4())[:8].upper()
        now = datetime.now()
        date_str = now.strftime("%d-%b-%Y")
        time_str = now.strftime("%I:%M %p").lstrip("0")

        print("\n---------------------------------------")
        print("      University Vending Machine")
        print(f"      Machine ID: {self.MACHINE_ID}")
        print(f"      Location: {self.LOCATION}")
        print("---------------------------------------")
        print(f"Transaction ID: {txn_id}")
        print(f"Date: {date_str}")
        print(f"Time: {time_str}")
        print("---------------------------------------")
        print("Item               Qty    Price")
        for it, q in self.cart.lines.items():
            print(f"{it.name:<18} {q:<6} {it.price * q:>8,} Ks")
        print("---------------------------------------")
        print(f"Total:           {total:>12,} Ks")
        print(f"Payment:         {paid:>12,} Ks")
        print(f"Change:          {change:>12,} Ks")
        print("---------------------------------------")
        print("Thank you for your purchase!")
        print("Contact: 09-123456789")
        print("---------------------------------------")

if __name__ == "__main__":
    VendingMachine().run()

File: VendingMachineApp.java
import java.util.*;
import java.time.*;
import java.time.format.DateTimeFormatter;

/**
 * Smart Vending Machine (User Flow Only) - Java Console
 * Features:
 *  - Browse items (drinks/snacks) with price & stock
 *  - Add to Cart (validates stock)
 *  - Checkout (cash only). Supports adding more cash or cancel.
 *  - Dispense simulation + Receipt printing
 *  - Stock auto-updates after successful purchase
 *
 * How to run:
 *   javac VendingMachineApp.java
 *   java VendingMachineApp
 */
public class VendingMachineApp {

    // --- Domain Models ---
    static class Item {
        final String name;
        final int price;    // kyats
        int stock;

        Item(String name, int price, int stock) {
            this.name = name;
            this.price = price;
            this.stock = stock;
        }
    }

    static class Cart {
        // keep insertion order for pretty receipt
        private final LinkedHashMap<Item, Integer> lines = new LinkedHashMap<>();

        void add(Item item, int qty) {
            lines.put(item, lines.getOrDefault(item, 0) + qty);
        }

        int total() {
            int sum = 0;
            for (Map.Entry<Item, Integer> e : lines.entrySet()) {
                sum += e.getKey().price * e.getValue();
            }
            return sum;
        }

        boolean isEmpty() {
            return lines.isEmpty();
        }

        Set<Map.Entry<Item, Integer>> entries() {
            return lines.entrySet();
        }

        void clear() {
            lines.clear();
        }
    }

    // --- App State ---
    private final Map<Integer, Item> inventory = new LinkedHashMap<>(); // menu number -> item
    private final Cart cart = new Cart();
    private final Scanner in = new Scanner(System.in);

    // Metadata for receipt
    private static final String MACHINE_ID = "VM-102";
    private static final String LOCATION   = "Building 209, Time City Complex";

    public static void main(String[] args) {
        new VendingMachineApp().run();
    }

    private void run() {
        seedInventory();
        printlnHeader();

        mainLoop:
        while (true) {
            showMenu();
            System.out.print("\nChoose: [number=add / C=Checkout / Q=Quit] : ");
            String choice = in.nextLine().trim();

            if (choice.equalsIgnoreCase("Q")) {
                System.out.println("\nGoodbye! Thanks for visiting.");
                break;
            }
            if (choice.equalsIgnoreCase("C")) {
                if (cart.isEmpty()) {
                    System.out.println("Cart is empty. Please add at least one item.");
                    continue;
                }
                if (checkout()) {
                    // after successful checkout, ask to continue shopping
                    System.out.print("\nBuy another item? (Y/N): ");
                    String again = in.nextLine().trim();
                    if (!again.equalsIgnoreCase("Y")) {
                        System.out.println("\nThank you! See you again.");
                        break mainLoop;
                    }
                }
                continue;
            }

            // Add to cart by number
            int num;
            try {
                num = Integer.parseInt(choice);
            } catch (NumberFormatException e) {
                System.out.println("Invalid choice. Try again.");
                continue;
            }
            Item item = inventory.get(num);
            if (item == null) {
                System.out.println("Invalid selection number.");
                continue;
            }
            if (item.stock <= 0) {
                System.out.println("Out of stock: " + item.name);
                continue;
            }

            int qty = askInt("Enter quantity for " + item.name + ": ", 1, item.stock);
            cart.add(item, qty);
            System.out.printf("Added to cart: %s x%d (%,d Ks each)%n", item.name, qty, item.price);

           System.out.print("Add more items? (Y to continue adding, anything else to return menu): ");
            String more = in.nextLine().trim();
            if (!more.equalsIgnoreCase("Y")) {
                // return to main menu
                continue;
            }

        }
    }

    // --- Checkout & Payment Flow ---
    private boolean checkout() {
        // Show cart summary
        System.out.println("\n================= CHECKOUT =================");
        printCart();
        int total = cart.total();
        System.out.printf("TOTAL: %,d Ks%n", total);

        // Confirm
        System.out.print("Proceed to payment? (Y/N): ");
        String ok = in.nextLine().trim();
        if (!ok.equalsIgnoreCase("Y")) {
            System.out.println("Returning to menu...");
            return false;
        }

        // Payment loop: allow adding more cash or cancel
        int paid = 0;
        while (paid < total) {
            System.out.printf("Insert cash (current paid %,d / total %,d): ", paid, total);
            String s = in.nextLine().trim();

            if (s.equalsIgnoreCase("CANCEL")) {
                System.out.println("Transaction canceled. Returning to main menu.");
                return false;
            }

            int add;
            try {
                add = Integer.parseInt(s);
                if (add <= 0) {
                    System.out.println("Please insert a positive amount (or type CANCEL).");
                    continue;
                }
            } catch (NumberFormatException e) {
                System.out.println("Invalid amount. Type a number or 'CANCEL'.");
                continue;
            }

            paid += add;
            if (paid < total) {
                System.out.printf("Insufficient funds. Need %,d Ks more.%n", (total - paid));
                System.out.print("Add more cash or type CANCEL to abort.\n");
            }
        }

        int change = paid - total;

        // Validate stock before finalizing (in case stock changed meanwhile)
        for (Map.Entry<Item, Integer> e : cart.entries()) {
            Item it = e.getKey();
            int q = e.getValue();
            if (it.stock < q) {
                System.out.printf("Sorry, '%s' stock changed. Only %d left. Transaction canceled.%n", it.name, it.stock);
                return false;
            }
        }

        // Dispense + Update Stock
        System.out.println("\nDispensing items...");
        for (Map.Entry<Item, Integer> e : cart.entries()) {
            Item it = e.getKey();
            int q = e.getValue();
            it.stock -= q;
            System.out.printf(" - %s x%d%n", it.name, q);
        }

        // Print Receipt
        printReceipt(paid, change);

        // Clear cart after purchase
        cart.clear();

        return true;
    }

    private void printCart() {
        System.out.println("-------------------------------------------");
        for (Map.Entry<Item, Integer> e : cart.entries()) {
            Item it = e.getKey();
            int q = e.getValue();
            System.out.printf("%-16s x%-3d  %,8d Ks%n", it.name, q, it.price * q);
        }
        System.out.println("-------------------------------------------");
    }

    private void printReceipt(int paid, int change) {
        int total = 0;
        for (Map.Entry<Item, Integer> e : cart.entries()) {
            total += e.getKey().price * e.getValue();
        }

        String txnId = genTxnId();
        String date = LocalDate.now().format(DateTimeFormatter.ofPattern("dd-MMM-uuuu"));
        String time = LocalTime.now().format(DateTimeFormatter.ofPattern("h:mm a"));
        System.out.println("\n---------------------------------------");
        System.out.println("      Gusto Vending Machine");
        System.out.println("      Machine ID: " + MACHINE_ID);
        System.out.println("      Location: " + LOCATION);
        System.out.println("---------------------------------------");
        System.out.println("Transaction ID: " + txnId);
        System.out.println("Date: " + date);
        System.out.println("Time: " + time);
        System.out.println("---------------------------------------");
        System.out.println("Item               Qty    Price");
        for (Map.Entry<Item, Integer> e : cart.entries()) {
            Item it = e.getKey();
            int q = e.getValue();
            System.out.printf("%-18s %-6d %,8d Ks%n", it.name, q, it.price * q);
        }
        System.out.println("---------------------------------------");
        System.out.printf("Total:           %,12d Ks%n", total);
        System.out.printf("Payment:         %,12d Ks%n", paid);
        System.out.printf("Change:          %,12d Ks%n", change);
        System.out.println("---------------------------------------");
        System.out.println("Thank you for your purchase!");
        System.out.println("Contact: 09-123456789");
        System.out.println("---------------------------------------");
    }

    private static String genTxnId() {
        // Short txn id from UUID
        return UUID.randomUUID().toString().substring(0, 8).toUpperCase();
    }

    // --- Menu & Utility ---
    private void showMenu() {
        System.out.println("\n=========== Available Items ===========");
        System.out.printf("%-4s %-18s %-10s %-8s%n", "No.", "Name", "Price(Ks)", "Stock");
        int i = 1;
        for (Item it : inventory.values()) {
            System.out.printf("%-4d %-18s %,10d %-8d%n", i, it.name, it.price, it.stock);
            i++;
        }
        System.out.println("=======================================");
        if (!cart.isEmpty()) {
            System.out.printf("Cart Total: %,d Ks%n", cart.total());
        }
        System.out.println("(Type item number to add, C=Checkout, Q=Quit)");
    }

private void printlnHeader() {
    // Use dividers for better visual separation
        System.out.println("-------------------------------------");

        // Center the main title and use capital letters for emphasis
        System.out.println("** Wellcome to GUSTO COLLEGE VENDING MACHINE **");
        
        // Use dividers for better visual separation
        System.out.println("------------------------------------------------------------------");
        
        // Display the warning message clearly with an icon
        System.out.println("Cash Payment Only | Type 'CANCEL' during payment to abort");
        
        System.out.println("------------------------------------------------------------------");
    }

    private void seedInventory() {
        // Drinks
        inventory.put(1, new Item("Coca-Cola", 2000, 10));
        inventory.put(2, new Item("Pepsi",     1800, 10));
        inventory.put(3, new Item("Sprite",    1500, 10));
        inventory.put(4, new Item("Water",     1000, 10));
        // Snacks
        inventory.put(5, new Item("Lays Chips", 1500, 8));
        inventory.put(6, new Item("Cookies",    1200, 6));
        inventory.put(7, new Item("Nutter Butter", 1300, 5));
        inventory.put(8, new Item("Kit Kat",    1500, 7));
    }

    private int askInt(String prompt, int min, int max) {
        while (true) {
            System.out.print(prompt);
            String s = in.nextLine().trim();
            try {
                int n = Integer.parseInt(s);
                if (n < min || n > max) {
                    System.out.printf("Please enter a number between %d and %d.%n", min, max);
                    continue;
                }
                return n;
            } catch (NumberFormatException e) {
                System.out.println("Invalid number. Try again.");
            }
        }
    }
}
